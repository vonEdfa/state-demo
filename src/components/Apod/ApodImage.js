import React from 'react';
import PropTypes from 'prop-types';
import Image from '../ImageLoaders/AspectRatioImageLoader';

const ApodImage = ({ title, url }) => {
  return (
    <Image
      src={url}
      alt={title}
      width="40%"
    />
  );
};

ApodImage.defaultProps = {
  title: '',
};

ApodImage.propTypes = {
  title: PropTypes.string,
  url: PropTypes.string.isRequired,
};

export default ApodImage;