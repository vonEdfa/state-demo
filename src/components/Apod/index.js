import React, { Component } from 'react';
import PropTypes from 'prop-types';

const timeZone = 'America/Los_Angeles';

class Apod extends Component {
  juan = 0;
  apodBaseUrl = 'https://api.nasa.gov/planetary/apod';
  key = process.env.REACT_APP_NASA_API_TOKEN;

  constructor(props) {
    super(props);
    this.state = {
      image: {
        url: '',
        hdurl: '',
        title: '',
        copyright: '',
        explanation: '',
        date: '',
        media_type: '',
      },
      selectedDate: this.toFormattedDate(this.props.date),
    };
  }

  toFormattedDate(date) {
    if (date instanceof Date) {
      return date.toLocaleDateString('se-sv');
    } else if (typeof date === 'string') {
      const pattern = /^[0-9]{4}-[0-9]{2}-[0-9]{2}/;
      if (pattern.test(date)) {
        return date;
      }
      try {
        const ddate = new Date(date);
        return this.toFormattedDate(ddate);;
      } catch (err) {
        console.error(err);
      }
    }
  }

  fetchImageData(date) {
    let image = {...this.state.image};
    console.debug('FETCHING', this.toFormattedDate(date));
    const formattedDate = this.toFormattedDate(date);
    let imageData = fetch(`${this.apodBaseUrl}?api_key=${this.key}&date=${formattedDate}`)
      .then(res => res.json())
      .then(res => {
        const { url, hdurl, title, copyright, explanation, date, media_type } = res;
        Object.assign(image, { url, hdurl, title, copyright, explanation, date, media_type });
        return image;
      })
      .catch(err => console.error(err));
      
    return imageData;
  }

  componentDidMount() {
    this.fetchImageData(this.props.date)
      .then((image) => {
        this.setState({ image,
          selectedDate: this.props.date });
      })
      .catch((err) => {
        console.error(err);
      });
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.selectedDate !== prevState.selectedDate)
      this.fetchImageData(this.state.selectedDate)
        .then(image => {
          this.setState({ image });
        });
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.date !== prevState.selectedDate) {
      return { selectedDate: nextProps.date };
    } else {
      return null;
    }
  }

  render() {
    return (
      <>
        <h1>APOD</h1>
        { this.state.image.media_type === 'image' ? (
          <img src={this.state.image.url} alt={this.state.image.title} />
        ) : (
          this.state.image.media_type
        ) }
      </>
    );
  }
}

Apod.defaultProps = {
  date: (new Date()).toLocaleDateString('se-sv', {timeZone: timeZone}),
};

Apod.propTypes = {
  date: PropTypes.oneOfType([
    PropTypes.instanceOf(Date),
  ])
};

export default Apod;