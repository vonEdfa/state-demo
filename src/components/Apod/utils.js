export const timeZone = 'America/Los_Angeles';

export const todayInCorrectTimeZone = (new Date()).toLocaleDateString('se-sv', {timeZone: timeZone});

export const toFormattedDate = date => {
  if (date instanceof Date) {
    return date.toLocaleDateString('se-sv');
  } else if (typeof date === 'string') {
    const pattern = /^[0-9]{4}-[0-9]{2}-[0-9]{2}/;
    if (pattern.test(date)) {
      return date;
    }
    try {
      const ddate = new Date(date);
      return toFormattedDate(ddate);;
    } catch (err) {
      console.error(err);
    }
  }
};