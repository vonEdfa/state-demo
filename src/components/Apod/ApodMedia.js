import React from 'react';
import PropTypes from 'prop-types';
import ApodImage from './ApodImage';

const ApodMedia = ({ media_type, title, url }) => {
  return (
    <>
      { media_type === 'image' ? (
        <ApodImage
          url={url}
          title={title}
        />
      ) : (
        media_type
      ) }
    </>
  );
};

ApodMedia.defaultProps = {
  media_type: 'image',
  title: '',
};

ApodMedia.propTypes = {
  media_type: PropTypes.string,
  title: PropTypes.string,
  url: PropTypes.string.isRequired,
};

export default ApodMedia;