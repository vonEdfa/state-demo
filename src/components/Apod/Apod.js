import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { toFormattedDate, todayInCorrectTimeZone } from './utils';

class Apod extends Component {
  apodBaseUrl = 'https://api.nasa.gov/planetary/apod';
  key = process.env.REACT_APP_NASA_API_TOKEN;

  constructor(props) {
    super(props);
    this.state = {
      image: {
        url: '',
        hdurl: '',
        title: '',
        copyright: '',
        explanation: '',
        date: '',
        media_type: '',
      },
      selectedDate: toFormattedDate(this.props.date),
    };
  }

  fetchImageData(date) {
    let image = {...this.state.image};
    console.debug('FETCHING', toFormattedDate(date));
    const formattedDate = toFormattedDate(date);
    let imageData = fetch(`${this.apodBaseUrl}?api_key=${this.key}&date=${formattedDate}`)
      .then(res => res.json())
      .then(res => {
        const { url, hdurl, title, copyright, explanation, date, media_type } = res;
        Object.assign(image, { url, hdurl, title, copyright, explanation, date, media_type });
        return image;
      })
      .catch(err => console.error(err));
      
    return imageData;
  }

  componentDidMount() {
    this.fetchImageData(this.props.date)
      .then((image) => {
        this.setState({ image,
          selectedDate: this.props.date });
      })
      .catch((err) => {
        console.error(err);
      });
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.selectedDate !== prevState.selectedDate)
      this.fetchImageData(this.state.selectedDate)
        .then(image => {
          this.setState({ image });
        });
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.date !== prevState.selectedDate) {
      return { selectedDate: nextProps.date };
    } else {
      return null;
    }
  }

  render() {
    return (
      <>
        <h1>APOD</h1>
        { this.state.image.media_type === 'image' ? (
          <img src={this.state.image.url} alt={this.state.image.title} />
        ) : (
          this.state.image.media_type
        ) }
      </>
    );
  }
}

Apod.defaultProps = {
  date: todayInCorrectTimeZone,
};

Apod.propTypes = {
  date: PropTypes.oneOfType([
    PropTypes.instanceOf(Date),
    PropTypes.string,
  ])
};

export default Apod;