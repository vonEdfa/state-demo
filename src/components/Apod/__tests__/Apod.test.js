import React from 'react';
import ReactDOM from 'react-dom';
import { act as domAct } from 'react-dom/test-utils';
import { act as testAct, create } from 'react-test-renderer';
import DateFnsUtils from '@date-io/date-fns';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import 'date-fns';
import Apod from '../Apod';
import ApodMedia from '../ApodMedia';
import ApodImage from '../ApodImage';

let container = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  ReactDOM.unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('Apod', () => {

  describe('Apod component', () => {

    it('renders without errors', () => {
      const errors = [];
      try {
        domAct(() => {
          ReactDOM.render(
            <MuiPickersUtilsProvider utils={DateFnsUtils}>
              <Apod />
            </MuiPickersUtilsProvider>,
            container
          );
        });
      } catch (error) {
        errors.push(error);
      }
      expect(errors).toStrictEqual([]);
    });

  });

  describe('ApodMedia component', () => {

    it('renders without errors', () => {
      const errors = [];
      try {
        domAct(() => {
          ReactDOM.render(
            <ApodMedia date="2019-10-31" />,
            container
          );
        });
      } catch (error) {
        errors.push(error);
      }
      expect(errors).toStrictEqual([]);
    });

  });

  describe('ApodImage component', () => {

    it('renders without errors', () => {
      const errors = [];
      try {
        domAct(() => {
          ReactDOM.render(
            <ApodImage url="https://apod.nasa.gov/apod/image/1910/ghostlyVeilNebula1034.jpg" />,
            container
          );
        });
      } catch (error) {
        errors.push(error);
      }
      expect(errors).toStrictEqual([]);
    });

    it('snapshot renders', () => {
      let root = null;
      domAct(() => {
        testAct(() => {
          root = create(<ApodImage url="https://apod.nasa.gov/apod/image/1910/ghostlyVeilNebula1034.jpg" />);
        });
      });
      expect(root).toMatchSnapshot();
    });

  });

});
