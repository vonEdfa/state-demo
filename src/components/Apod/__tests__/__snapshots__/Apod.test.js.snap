// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apod ApodImage component snapshot renders 1`] = `
<div
  style={
    Object {
      "width": "40%",
    }
  }
>
  <div
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.2)",
        "height": 0,
        "overflow": "hidden",
        "paddingTop": "80%",
        "position": "relative",
        "transition": "0.5s ease-in-out",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "display": "flex",
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <div
        className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style={
          Object {
            "height": 40,
            "position": "absolute",
            "width": 40,
            "zIndex": 1000,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </div>
      <img
        alt="Loading..."
        src=""
        style={
          Object {
            "fontSize": "0.5em",
            "height": "auto",
            "opacity": 0.5,
            "transition": "1s ease-in-out",
            "width": "100%",
          }
        }
      />
    </div>
  </div>
</div>
`;
